File: ts/kip-do/src/schema/package.ts
import { parseCanonicalJson } from '@ldclabs/kip-lang'
/**
 * # The Schema Package artifact
 *
 * Authoritative Schema in KIP 2.0 is an immutable versioned artifact, not a set
 * of graph nodes (Spec §20.1, §20.11). This is the difference from 1.x that costs
 * the most to get wrong: in 1.x an ordinary `UPSERT` could reshape what a type
 * meant, so cognition and the rules governing cognition sat in the same mutable
 * store. Here a package is content-addressed and read-only, and changing
 * semantics means publishing a new version and activating it through
 * Governance.
 *
 * The types below mirror the shipped artifact format — the same JSON as
 * `KIP/v2/profiles/cognitive-memory-2.0.0.schema.json`, which the tests parse
 * rather than a hand-written imitation of it.
 *
 * Every member is optional and unknown members are kept: a package written
 * against a later minor format revision must stay readable, and an engine that
 * dropped the members it did not recognize would silently change the artifact's
 * digest-covered content.
 */

import { errors } from '../errors.js'
import { isJsonMap, type Json, type JsonMap } from '../json.js'
import {
  parsePackageRef,
  SECTIONS,
  type PackageRef,
  type SymbolKind,
  type SymbolRef,
} from './symbol.js'

/** One declared field of an attribute set or a Facet. */
export interface FieldSpec {
  /** The validation type: a name, or a list of accepted names. */
  type?: Json
  /** Whether the field must be present (§36). */
  required?: boolean
  /** Whether the field may change after creation (§39). Defaults to true. */
  mutable?: boolean
  minimum?: number
  maximum?: number
  /** The accepted values, when the field is a closed set. */
  enum?: Json[]
  /** The value used when the field is absent (§40). */
  default?: Json
  [extra: string]: Json | undefined
}

/** The attribute contract of a Concept type (§34–§40). */
export interface AttributeSpec {
  value_schema?: Json
  /** Whether attributes not named here are permitted (§37). */
  open?: boolean
  fields?: Record<string, FieldSpec>
}

/** What may occupy one end of a Proposition or structural edge. */
export interface EndpointSpec {
  /** Permitted Concept types, as canonical references. */
  concept_types?: string[]
  /** Permitted Core element kinds, e.g. `Concept`, `Assertion`. */
  kinds?: string[]
  /**
   * Permitted Literal datatypes, when a Literal is allowed (§20.15).
   *
   * §20.15 spells the member `literal_types` and draws its values from §9.2's
   * four names; `datatypes` is the older spelling, read the same way.
   */
  literal_types?: string[]
  datatypes?: string[]
  /** Whether `null` is a permitted object (§9.5, §20.15). */
  nullable?: boolean
  /**
   * The shape a string Literal must have — `timestamp`, `uri`, or a
   * package-defined name (§20.15). Validated on write, never part of identity.
   */
  format?: string
  [extra: string]: Json | undefined
}

/** The Literal datatypes an endpoint admits, whichever member spells them. */
export function literalTypesOf(spec: EndpointSpec | undefined): string[] {
  return spec?.literal_types ?? spec?.datatypes ?? []
}

/** Whether an endpoint declares no restriction at all. */
export function isUnconstrained(spec: EndpointSpec | undefined): boolean {
  return (
    (spec?.concept_types?.length ?? 0) === 0 &&
    (spec?.kinds?.length ?? 0) === 0 &&
    literalTypesOf(spec).length === 0
  )
}

/**
 * A Concept type definition (§32).
 *
 * A Concept type says what a Concept *is*, never whether anything about it is
 * true (§33).
 */
export interface ConceptTypeDef {
  ref?: string
  kind?: string
  description?: string
  attributes?: AttributeSpec
  /**
   * Anything a later format revision added.
   *
   * Typed as `unknown` rather than `Json` because the declared members above
   * are structured: an index signature has to admit every member's type, and
   * narrowing it to `Json` would make the structured ones unassignable.
   */
  [extra: string]: unknown
}

/** A predicate type definition (§41). */
export interface PredicateDef {
  ref?: string
  kind?: string
  /** What the predicate means. Semantics must be explicit (§57). */
  description?: string
  subject?: EndpointSpec
  object?: EndpointSpec
  /**
   * Whether one subject has at most one accepted object at one valid time
   * (§20.15, §25.1).
   *
   * This is an *epistemic* statement, not a storage constraint: a functional
   * predicate with two competing objects is a conflict set, which the engine
   * must be able to store in order to report it (§25, §95).
   */
  functional?: boolean
  /**
   * Whether absence of a Proposition means insufficient (§24) rather than a
   * closed-world absence (§24.2). Defaults to `true`.
   */
  open_world?: boolean
  /**
   * Whether the candidate objects of a functional slot are exclusive:
   * accepting one rejects the others (§20.15, §25). Defaults to `false`.
   */
  complete?: boolean
  /**
   * Whether, for a boolean-valued Predicate, object `false` is the negation
   * of object `true` (§12.7, §20.15). Defaults to `false`.
   */
  boolean_completeness?: boolean
  /**
   * When two accepted values of a functional slot conflict (§25.2):
   * `overlapping_valid_time` (the default) or `none`.
   */
  temporal_conflict?: string
  /**
   * Anything a later format revision added.
   *
   * Typed as `unknown` rather than `Json` because the declared members above
   * are structured: an index signature has to admit every member's type, and
   * narrowing it to `Json` would make the structured ones unassignable.
   */
  [extra: string]: unknown
}

/**
 * A Facet definition (§58).
 *
 * A Facet is a validated namespaced extension, not an untyped metadata bag
 * (§18.1) — which is exactly what KIP 1.x `metadata` had become.
 */
export interface FacetDef {
  ref?: string
  kind?: string
  description?: string
  /** Whether members not named here are rejected (§60). */
  closed?: boolean
  applicable_to?: EndpointSpec
  fields?: Record<string, FieldSpec>
  /**
   * Anything a later format revision added.
   *
   * Typed as `unknown` rather than `Json` because the declared members above
   * are structured: an index signature has to admit every member's type, and
   * narrowing it to `Json` would make the structured ones unassignable.
   */
  [extra: string]: unknown
}

/** A structural field's permitted reference count. */
export interface Cardinality {
  min?: number
  /** The maximum; absent means unbounded. */
  max?: number | null
}

/**
 * A structural field definition (§62).
 *
 * Structural fields are record topology, not semantic Propositions (§64): a
 * claim *about* a structural relation is a separate Proposition plus Assertion.
 */
export interface StructuralFieldDef {
  ref?: string
  kind?: string
  description?: string
  source?: EndpointSpec
  target?: EndpointSpec
  cardinality?: Cardinality
  /**
   * Whether edge order is meaningful (§66).
   *
   * Order is not causality: `has_step` being ordered says step 3 follows step
   * 2, never that it was caused by it.
   */
  ordered?: boolean
  unique?: boolean
  /**
   * Anything a later format revision added.
   *
   * Typed as `unknown` rather than `Json` because the declared members above
   * are structured: an index signature has to admit every member's type, and
   * narrowing it to `Json` would make the structured ones unassignable.
   */
  [extra: string]: unknown
}

/** The symbols a package defines, one map per symbol kind. */
export interface Definitions {
  concept_types?: Record<string, ConceptTypeDef>
  predicates?: Record<string, PredicateDef>
  facets?: Record<string, FacetDef>
  structural_fields?: Record<string, StructuralFieldDef>
  enums?: Record<string, Json>
  /** Additions to Core's open registries, e.g. `activity_classes` (§69). */
  registry_extensions?: Record<string, Json>
}

/** Identity and provenance. */
export interface Manifest {
  package_id?: string
  version?: string
  /** The two above, joined: `kip://core@2.0.0`. */
  package_ref?: string
  name?: string
  description?: string
  /** Who published it. Namespace identity does not prove this (§20.11). */
  publisher?: Json
  [extra: string]: Json | undefined
}

/** One resolved dependency. */
export interface Dependency {
  package_id?: string
  version?: string
  package_ref?: string
  /** Whether activation fails without it (§73). */
  required?: boolean
  [extra: string]: Json | undefined
}

/** The content digest and signatures. */
export interface Integrity {
  digest_profile?: string
  content_digest?: string
  covers?: string
  /** Signatures over the digest. A signature is not local approval (§90). */
  signatures?: Json[]
}

/** A published Schema Package. */
export interface SchemaPackage {
  format?: string
  format_version?: string
  manifest?: Manifest
  dependencies?: Dependency[]
  definitions?: Definitions
  constraints?: Json[]
  aliases?: JsonMap
  compatibility?: Json
  migrations?: Json
  /** Advisory guidance for an Agent. Never a validator (§20.5). */
  model_hints?: Json
  canonicalization?: Json
  integrity?: Integrity
  /**
   * Anything a later format revision added.
   *
   * Typed as `unknown` rather than `Json` because the declared members above
   * are structured: an index signature has to admit every member's type, and
   * narrowing it to `Json` would make the structured ones unassignable.
   */
  [extra: string]: unknown
}

/** Parses an artifact and checks the identity it declares is coherent. */
export function parsePackage(source: string | JsonMap): SchemaPackage {
  let value: unknown
  if (typeof source === 'string') {
    try {
      value = parseCanonicalJson(source)
    } catch (err) {
      throw errors.artifactParseError(
        `this is not a readable Schema Package artifact: ${String(err)}`,
      )
    }
  } else {
    value = source
  }
  if (!isJsonMap(value)) {
    throw errors.artifactParseError(
      'a Schema Package artifact must be a JSON object',
    )
  }
  const artifact = value as SchemaPackage
  // Reject here rather than at the first symbol lookup: an artifact whose two
  // spellings of its own identity disagree could be installed under a name its
  // symbols do not claim.
  packageRefOf(artifact)
  return artifact
}

/**
 * The package's exact identity.
 *
 * `package_ref` is the authority when present, and `package_id@version` must
 * agree with it.
 */
export function packageRefOf(artifact: SchemaPackage): PackageRef {
  const manifest = artifact.manifest ?? {}
  const joined = `${manifest.package_id ?? ''}@${manifest.version ?? ''}`
  const declared =
    manifest.package_ref === undefined || manifest.package_ref === ''
      ? joined
      : manifest.package_ref
  if (
    (manifest.package_id ?? '') !== '' &&
    (manifest.version ?? '') !== '' &&
    declared !== joined
  ) {
    throw errors.capsuleValidationFailed(
      `the artifact calls itself ${JSON.stringify(declared)} but its ` +
        `package_id and version join to ${JSON.stringify(joined)}`,
    )
  }
  return parsePackageRef(declared)
}

function section(
  artifact: SchemaPackage,
  kind: SymbolKind,
): Record<string, unknown> {
  const definitions = artifact.definitions ?? {}
  return (
    (definitions[SECTIONS[kind] as keyof Definitions] as
      | Record<string, unknown>
      | undefined) ?? {}
  )
}

/** Whether this package defines a symbol of the given kind. */
export function defines(
  artifact: SchemaPackage,
  kind: SymbolKind,
  name: string,
): boolean {
  return Object.hasOwn(section(artifact, kind), name)
}

/** The local names this package defines for one symbol kind. */
export function symbols(
  artifact: SchemaPackage,
  kind: SymbolKind,
): string[] {
  return Object.keys(section(artifact, kind))
}

/**
 * The Core element kinds `kip://core` exports (§20.13).
 *
 * Mirrors `anda_kip::CORE_ELEMENT_KINDS`.
 */
export const CORE_ELEMENT_KINDS: readonly string[] = [
  'Concept',
  'Proposition',
  'Assertion',
  'Evidence',
  'Activity',
]

/**
 * The reserved Core structural fields `kip://core` exports, each with the Core
 * kind that owns it (§20.13).
 *
 * Mirrors `anda_kip::CORE_STRUCTURAL_FIELDS`. Resolved by the source element's
 * Core kind, never through a package alias — which is why a package field of
 * the same name, carried by the *same* kind, would change what an Assertion
 * cites without changing any command that reads it. On a different kind there
 * is nothing to shadow: a Concept owns no Core structural field, so a Profile
 * `evidence` on a Concept is a separate plane.
 */
export const CORE_STRUCTURAL_FIELD_OWNERS: Readonly<Record<string, string>> = {
  evidence: 'Assertion',
  context: 'Assertion',
  source: 'Evidence',
  generated_by: 'Evidence',
  inputs: 'Activity',
  outputs: 'Activity',
  associated_actors: 'Activity',
}

/**
 * Refuses a package that would shadow a reserved Core symbol (§20.13).
 *
 * `kip://core` is implicitly active in every Schema Environment and cannot be
 * deactivated, replaced, or shadowed. A package defining a Concept type named
 * `Assertion`, or an Activity-sourced structural field named `inputs`, would
 * put two meanings behind one word in one resolution scope — and the reader
 * that resolved the wrong one could not tell.
 *
 * The namespaces are checked apart. A type, Facet or Enum resolves by name
 * alone, so a Core element kind is a shadow wherever it appears. A structural
 * field resolves by source kind *and* name (§8.2), so it shadows only where
 * the kind that owns the Core field could carry it — a source constraining
 * nothing admits every kind, and therefore shadows. A Predicate named `source`
 * is a claim about origin and shadows nothing.
 */
export function rejectCoreShadowing(artifact: SchemaPackage): void {
  const refuse = (kind: SymbolKind, name: string, detail: string): never => {
    throw errors.constraintViolation(
      `this package defines the ${SECTIONS[kind]} \`${name}\`, which is a ` +
        `reserved Core symbol (§20.13)${detail}: \`kip://core\` is implicitly ` +
        `active in every Schema Environment and cannot be shadowed. Rename it, ` +
        `or address the Core symbol you meant`,
    )
  }

  for (const kind of ['ConceptType', 'Facet', 'Enum'] as const) {
    for (const name of symbols(artifact, kind)) {
      if (CORE_ELEMENT_KINDS.includes(name)) refuse(kind, name, '')
    }
  }

  const fields = artifact.definitions?.structural_fields ?? {}
  for (const [name, def] of Object.entries(fields)) {
    const owner = CORE_STRUCTURAL_FIELD_OWNERS[name]
    if (owner === undefined) continue
    const kinds = def.source?.kinds ?? []
    const types = def.source?.concept_types ?? []
    if (kinds.length === 0 && types.length === 0) {
      refuse(
        'StructuralField',
        name,
        ` on ${owner}: a source that constrains nothing admits every kind, ${owner} included, and ${owner}'s \`${name}\` is defined by the protocol itself`,
      )
    }
    if (kinds.includes(owner)) {
      refuse(
        'StructuralField',
        name,
        ` on ${owner}, whose \`${name}\` the protocol itself defines`,
      )
    }
  }
}

/** The canonical reference for one of this package's local names. */
export function symbolRefOf(
  artifact: SchemaPackage,
  name: string,
): SymbolRef {
  return { package: packageRefOf(artifact), name }
}

export const conceptTypeDef = (
  artifact: SchemaPackage,
  name: string,
): ConceptTypeDef | undefined => artifact.definitions?.concept_types?.[name]

export const predicateDef = (
  artifact: SchemaPackage,
  name: string,
): PredicateDef | undefined => artifact.definitions?.predicates?.[name]

/**
 * A Predicate definition's declarations with §20.15's defaults filled in.
 *
 * `functional: false`, `open_world: true`, `complete: false`,
 * `boolean_completeness: false`, `temporal_conflict: "overlapping_valid_time"`.
 * An absent definition — a predicate this environment cannot resolve —
 * declares nothing, and reads as the defaults too.
 */
export interface PredicateRules {
  functional: boolean
  open_world: boolean
  complete: boolean
  boolean_completeness: boolean
  temporal_conflict: 'overlapping_valid_time' | 'none'
}

export function predicateRules(def: PredicateDef | undefined): PredicateRules {
  return {
    functional: def?.functional === true,
    open_world: def?.open_world !== false,
    complete: def?.complete === true,
    boolean_completeness: def?.boolean_completeness === true,
    temporal_conflict: def?.temporal_conflict === 'none' ? 'none' : 'overlapping_valid_time',
  }
}

export const facetDef = (
  artifact: SchemaPackage,
  name: string,
): FacetDef | undefined => artifact.definitions?.facets?.[name]

export const structuralFieldDef = (
  artifact: SchemaPackage,
  name: string,
): StructuralFieldDef | undefined =>
  artifact.definitions?.structural_fields?.[name]

File: rs/anda_cognitive_nexus/src/migrate/package.rs
//! Phase 3a: a Schema Package for the vocabulary 1.x invented at runtime.
//!
//! In 1.x a type was a string on a row, so a deployment's vocabulary is
//! whatever its writers happened to use. In 2.0 every `schema_ref` names an
//! exact symbol in an immutable versioned Package (§20.4), which leaves a
//! migration two options: discard the legacy types, or publish a Package that
//! contains them.
//!
//! This publishes one, per the guide's §8:
//!
//! ```text
//! kip://legacy/<deployment>@1.0.0
//! ```
//!
//! The vocabulary is read off the data rather than declared, because the 1.x
//! `$ConceptType` graph nodes were advisory — a 1.x write could use a type
//! nobody had declared, and refusing those rows at migration time would lose
//! exactly the records that most need explaining.
//!
//! Two things it does not do. It does not put legacy types into
//! `kip://core` — Core is the protocol's, not a deployment's. And it does not
//! constrain: every generated type is open-attribute and every predicate is
//! non-functional and open-world, because a constraint invented here would be
//! one the 1.x data was never checked against, and the first thing it would do
//! is reject the deployment's own history.

use anda_kip::{Json, KipError, Map};
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::collections::{BTreeMap, BTreeSet};

use super::stage::LegacyRow;
use crate::schema::{Intent, SchemaEnvironment, SymbolKind};

/// The package id every migrated symbol resolves through.
pub const LEGACY_PACKAGE_ID: &str = "kip://legacy/nexus";
/// Its version. One migration, one version.
pub const LEGACY_PACKAGE_VERSION: &str = "1.1.0";

/// The exact reference migrated elements are written against.
pub fn legacy_package_ref() -> String {
    format!("{LEGACY_PACKAGE_ID}@{LEGACY_PACKAGE_VERSION}")
}

fn symbol(name: &str) -> String {
    format!("{}/{}", legacy_package_ref(), name)
}

/// The vocabulary a set of staged rows actually uses.
///
/// Split in two after [`Vocabulary::resolve`]: names the Space can already
/// resolve are *adopted* — a migrated element is written against the host's own
/// symbol — and only what is left is declared here.
///
/// That split is the whole point. A 1.x Brain used `Person` and `Event`, and so
/// does the Cognitive Memory Profile a 2.0 host activates.
/// Minting a legacy `Person` beside the profile's would leave two symbols
/// spelled the same, and every command naming the bare local name would resolve
/// to neither — `SchemaSymbolAmbiguous`, on a Space whose data migrated
/// perfectly.
#[derive(Debug, Default, Serialize, Deserialize)]
pub struct Vocabulary {
    #[serde(default = "legacy_package_ref")]
    package_ref: String,
    /// Every distinct 1.x Concept `type` this package declares.
    pub concept_types: BTreeSet<String>,
    /// Every distinct 1.x predicate this package declares.
    pub predicates: BTreeSet<String>,
    /// Legacy name → the exact symbol an already-active package provides.
    adopted_types: BTreeMap<String, String>,
    adopted_predicates: BTreeMap<String, String>,
    #[serde(default)]
    predicate_overrides: BTreeMap<String, String>,
}

impl Vocabulary {
    pub fn package_ref(&self) -> String {
        if self.package_ref.is_empty() {
            legacy_package_ref()
        } else {
            self.package_ref.clone()
        }
    }

    /// Store exact references for every source symbol, including the generated
    /// ones, so a later binary need not re-render them from its own constants.
    pub fn freeze_refs(&mut self) {
        self.package_ref = self.package_ref();
        for name in &self.concept_types {
            self.adopted_types
                .entry(name.clone())
                .or_insert_with(|| symbol(name));
        }
        for name in &self.predicates {
            self.adopted_predicates
                .entry(name.clone())
                .or_insert_with(|| symbol(name));
        }
    }
    /// Reads the vocabulary off the staged rows.
    pub fn scan(concepts: &[LegacyRow], propositions: &[LegacyRow]) -> Self {
        let mut vocabulary = Vocabulary::default();
        for row in concepts {
            if let Some(name) = row.doc.get("type").and_then(Json::as_str)
                && !name.is_empty()
            {
                vocabulary.concept_types.insert(name.to_string());
            }
        }
        for row in propositions {
            // 1.x stored a *set* of predicates per row, each with its own
            // properties: one row is many 2.0 tuples, so every member counts.
            if let Some(list) = row.doc.get("predicates").and_then(Json::as_array) {
                for predicate in list.iter().filter_map(Json::as_str) {
                    if !predicate.is_empty() {
                        vocabulary.predicates.insert(predicate.to_string());
                    }
                }
            }
        }
        vocabulary
    }

    /// Hands every name the Space can already resolve to the package that
    /// provides it, and keeps the rest for this one.
    ///
    /// Adoption is by local name, which is the same identity 1.x had: a 1.x
    /// `Person` and the profile's `Person` are the same word for the same
    /// thing, and the guide's §8 says exactly this — migrate toward the
    /// standard profile where the semantics match.
    ///
    /// Where they do not match, nothing is adopted: a name no active package
    /// declares stays here, unconstrained, because a 1.x deployment's
    /// `ShipmentLeg` means whatever that deployment meant by it.
    pub fn resolve(&mut self, env: &SchemaEnvironment) {
        let mut reserved_types = self.concept_types.clone();
        for (kind, names, adopted) in [
            (
                SymbolKind::ConceptType,
                &mut self.concept_types,
                &mut self.adopted_types,
            ),
            (
                SymbolKind::PredicateType,
                &mut self.predicates,
                &mut self.adopted_predicates,
            ),
        ] {
            let mut keep = BTreeSet::new();
            for name in std::mem::take(names) {
                // `Intent::Write`: a migrated element is written against this
                // symbol, so a package that may be read but not written to is
                // not one to adopt.
                match env.resolve_symbol(kind, &name, Intent::Write) {
                    Ok(reference) if reference.package.package_id != LEGACY_PACKAGE_ID => {
                        // A legacy learned/runtime artifact cannot acquire v2 standing
                        // merely because it used the same type name. Keep its complete
                        // content under a distinct, open legacy type instead.
                        if kind == SymbolKind::ConceptType
                            && reference.package.package_id == crate::profiles::COGNITIVE_MEMORY_ID
                            && !["Person", "Event", "Insight", "Commitment", "SleepTask"]
                                .contains(&name.as_str())
                        {
                            let base = format!("Legacy{name}");
                            let mut legacy_name = base.clone();
                            let mut suffix = 2;
                            while reserved_types.contains(&legacy_name)
                                || env
                                    .resolve_symbol(
                                        SymbolKind::ConceptType,
                                        &legacy_name,
                                        Intent::Read,
                                    )
                                    .is_ok()
                            {
                                legacy_name = format!("{base}{suffix}");
                                suffix += 1;
                            }
                            reserved_types.insert(legacy_name.clone());
                            adopted.insert(name, symbol(&legacy_name));
                            keep.insert(legacy_name);
                        } else {
                            adopted.insert(name, reference.to_string());
                        }
                    }
                    _ => {
                        keep.insert(name);
                    }
                }
            }
            *names = keep;
        }
    }

    /// Whether there is anything to publish.
    pub fn is_empty(&self) -> bool {
        self.concept_types.is_empty() && self.predicates.is_empty()
    }

    /// The exact symbol a legacy Concept type resolves to.
    pub fn concept_ref(&self, name: &str) -> Option<String> {
        self.symbol_ref(SymbolKind::ConceptType, &self.concept_types, name)
    }

    /// The exact symbol a legacy predicate resolves to.
    pub fn predicate_ref(&self, name: &str) -> Option<String> {
        self.symbol_ref(SymbolKind::PredicateType, &self.predicates, name)
    }

    /// A v1 predicate could be used with endpoints a native package rejects.
    /// Preserve just those tuples under a distinct legacy predicate; keep
    /// correctly typed tuples on the host vocabulary used by normal recall.
    pub fn resolve_legacy_endpoints(
        &mut self,
        env: &SchemaEnvironment,
        concepts: &[LegacyRow],
        propositions: &[LegacyRow],
    ) {
        use crate::schema::EndpointFacts;
        use anda_kip::ElementKind;
        let types: BTreeMap<_, _> = concepts
            .iter()
            .filter_map(|row| Some((row.legacy_id, self.concept_ref(row.doc["type"].as_str()?)?)))
            .collect();
        let endpoint = |value: &Json| -> EndpointFacts {
            let reference = value.as_str().unwrap_or("");
            if let Some(id) = reference
                .strip_prefix("C:")
                .and_then(|s| s.parse::<u64>().ok())
            {
                return EndpointFacts::Element {
                    kind: ElementKind::Concept,
                    schema_ref: types.get(&id).cloned(),
                };
            }
            if reference.starts_with("P:") {
                return EndpointFacts::Element {
                    kind: ElementKind::Proposition,
                    schema_ref: None,
                };
            }
            EndpointFacts::Unresolved
        };
        let mut aliases = BTreeMap::new();
        let mut reserved: BTreeSet<_> = self
            .predicates
            .iter()
            .chain(self.adopted_predicates.keys())
            .cloned()
            .collect();
        for row in propositions {
            for name in row.doc["predicates"]
                .as_array()
                .into_iter()
                .flatten()
                .filter_map(Json::as_str)
            {
                let Some(reference) = self.adopted_predicates.get(name) else {
                    continue;
                };
                let valid = env
                    .prepare_proposition(
                        reference,
                        &endpoint(&row.doc["subject"]),
                        &endpoint(&row.doc["object"]),
                        Intent::Write,
                    )
                    .is_ok_and(|(_, v)| v.is_valid());
                if valid {
                    continue;
                }
                let reference = aliases.entry(name.to_string()).or_insert_with(|| {
                    let base = format!("legacy_{name}");
                    let mut alias = base.clone();
                    let mut suffix = 2;
                    while reserved.contains(&alias)
                        || env
                            .resolve_symbol(SymbolKind::PredicateType, &alias, Intent::Read)
                            .is_ok()
                    {
                        alias = format!("{base}_{suffix}");
                        suffix += 1;
                    }
                    reserved.insert(alias.clone());
                    self.predicates.insert(alias.clone());
                    symbol(&alias)
                });
                self.predicate_overrides
                    .insert(format!("P:{}:{name}", row.legacy_id), reference.clone());
            }
        }
    }

    pub fn predicate_for(&self, legacy_id: u64, name: &str) -> Option<String> {
        self.predicate_overrides
            .get(&format!("P:{legacy_id}:{name}"))
            .cloned()
            .or_else(|| self.predicate_ref(name))
    }

    fn symbol_ref(
        &self,
        kind: SymbolKind,
        declared: &BTreeSet<String>,
        name: &str,
    ) -> Option<String> {
        let adopted = match kind {
            SymbolKind::ConceptType => &self.adopted_types,
            _ => &self.adopted_predicates,
        };
        if let Some(adopted) = adopted.get(name) {
            return Some(adopted.clone());
        }
        declared.contains(name).then(|| symbol(name))
    }

    /// Renders the Package artifact.
    pub fn artifact(&self) -> Result<Json, KipError> {
        let mut concept_types = Map::new();
        for name in &self.concept_types {
            concept_types.insert(
                name.clone(),
                json!({
                    "ref": symbol(name),
                    "kind": "ConceptType",
                    "description": format!(
                        "Migrated KIP 1.x Concept type {name:?}. Generated from the data, \
                         not from a declaration: 1.x types were strings on rows."
                    ),
                    // Open, and with no declared field: 1.x attributes were
                    // free-form, and a field list derived from today's rows
                    // would silently forbid tomorrow's.
                    "attributes": {"open": true, "fields": {}},
                }),
            );
        }

        let mut predicates = Map::new();
        for name in &self.predicates {
            predicates.insert(
                name.clone(),
                json!({
                    "ref": symbol(name),
                    "kind": "PredicateType",
                    "description": format!(
                        "Migrated KIP 1.x predicate {name:?}. Generated from the data."
                    ),
                    // Unconstrained ends: 1.x never checked them, so a domain
                    // or range asserted here would be a claim about the old
                    // system that the old system never enforced.
                    "subject": {"kinds": ["Concept", "Proposition"]},
                    "object": {"kinds": ["Concept", "Proposition", "Literal"]},
                    // Never functional. Marking a legacy predicate functional
                    // would turn two rows that coexisted for years into a
                    // conflict the projection has to resolve.
                    "functional": false,
                    "open_world": true,
                    "complete": false,
                }),
            );
        }

        Ok(json!({
            "format": "KIP-Schema-Package",
            "format_version": "2.0",
            "manifest": {
                "package_id": LEGACY_PACKAGE_ID,
                "version": LEGACY_PACKAGE_VERSION,
                "package_ref": legacy_package_ref(),
                "name": "KIP 1.x legacy vocabulary",
                "description":
                    "Concept types and predicates recovered from a KIP 1.x deployment during \
                     migration. Compatibility surface, not a designed ontology: migrate toward \
                     kip://profiles/cognitive-memory where the semantics actually match.",
                "publisher": "urn:kip:publisher:migration",
                "purpose": "compatibility",
            },
            "dependencies": [],
            "definitions": {
                "concept_types": concept_types,
                "predicates": predicates,
                "facets": {
                    "LegacyRecord": {
                        "ref": symbol("LegacyRecord"),
                        "kind": "FacetDefinition",
                        "description": "Verbatim KIP 1.x source row. Migration provenance, not independent Evidence or Governance authority.",
                        "closed": true,
                        "applicable_to": {"kinds": ["Concept", "Assertion"]},
                        "fields": {"record": {"type": "object", "required": true, "mutable": false}}
                    }
                },
                "structural_fields": {},
                "enums": {},
                "registry_extensions": {},
            },
            "constraints": [],
            "aliases": {},
            "compatibility": {},
            "migrations": [],
            "model_hints": {},
        }))
    }
}
